### PR TITLE
Add support for batching through a SQL backend

### PR DESCRIPTION
**Pull request by RickHawesUSDS**
This PR does adds batching of reports. In addition, it also creates a DB backend to track and audit reports as they flow through the system. 

## Developer Note
- You need a local prime_data_hub database (empty is fine) to build. See the `getting_started.md` instructions. 
- To deploy to you dev resource groups you need to setup a database. Run `setup_database.sh` for a script. 

## Changes
- DataAccess, QueueAccess and BlobAccess data access layer 
- Batch Function - Process Function renamed to Translate 
- Hikari DB connection Pools 
- Batch definition in organizations.yml 
- Setup_database.sh and Dockerfile improvements

## Checklist
- [x] Tested locally?
- [x] Added new dependencies?
- [x] Updated the docs?
- [x] Added a test?
- [x] Did you check for sensitive data, and remove any?
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

## Fixes 

- #72 Get Logging Working
- #66 Aggregate messages
- #65 Timing of batches
- #30 Router as cloud service

## To Be Done

- #issue

### CODE REVIEW

File: prime-router/Dockerfile [hunk 1]
```diff
@@ -9,7 +9,6 @@ FROM mcr.microsoft.com/azure-functions/java:3.0-java$JAVA_VERSION-build AS insta
 COPY . /src/java-function-app
 RUN cd /src/java-function-app && \
     mkdir -p /home/site/wwwroot && \
-    mvn clean package && \
     cd ./target/azure-functions/ && \
```
Comment (RickHawesUSDS) [inline prime-router/Dockerfile:12]: I'm removing the Java build step for two reasons: 1) much faster to build the image 2) difficulty connecting to local database.

File: prime-router/docker-compose.yml [hunk 2]
```diff
@@ -1,21 +1,23 @@
...
         target: /prime-data-hub-router
     entrypoint: [/prime-data-hub-router/start_func.sh]
     environment: 
-      AzureWebJobsStorage: DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;
-    depends_on: [azurite]   
+      - AzureWebJobsStorage=DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;
+      - POSTGRES_USER=prime
+      - POSTGRES_PASSWORD=changeIT!
+      - POSTGRES_URL=jdbc:postgresql://host.docker.internal:5432/prime_data_hub
+    depends_on: [azurite]
```
Comment (RickHawesUSDS) [inline prime-router/docker-compose.yml:20]: POSTGRES environment variables will contain the DB connection string.

File: prime-router/metadata/organizations.yml [hunk 3]
```diff
@@ -19,6 +19,11 @@
         jurisdictionalFilter: { standard.testing_lab_state: AZ }
 # OLD        jurisdictionalFilter: { standard.patient_state: AZ }
         transforms: { deidentify: false }
+        batch:
+          operation: MERGE
+          numberPerDay: 1440 # Every minute
+          initialBatch: 00:00
+          timeZone: ARIZONA
         address: http://localhost:1181/
```
Comment (RickHawesUSDS) [inline prime-router/metadata/organizations.yml:27]: For now, while we are debugging, send every minute. Will change once we onboard and a regular cadence is established.

File: prime-router/pom.xml [hunk 4]
```diff
@@ -251,6 +259,65 @@
...
+                    </execution>
+                </executions>
+                <configuration>
+                    <jdbc>
+                        <driver>org.postgresql.Driver</driver>
+                        <url>jdbc:postgresql://127.0.0.1:5432/prime_data_hub</url>
+                        <user>prime</user>
+                        <password>changeIT!</password>
+                    </jdbc>
+                    <generator>
```
Comment (RickHawesUSDS) [inline prime-router/pom.xml:301]: Jooq requires a local database to extract the schema from.

File: prime-router/setup_database.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+#
+# Setup a dev database in a developer resource group
+#
+
```
Comment (RickHawesUSDS) [inline prime-router/setup_database.sh:5]: Meant to be run after setup_resource_group.sh

File: prime-router/src/main/kotlin/Hl7Converter.kt [hunk 6]
```diff
@@ -182,7 +182,7 @@ object Hl7Converter {
     }
 
     private fun createBTS(report: Report): String {
-        return "BTS|${report.rowCount}\r"
+        return "BTS|${report.itemCount}\r"
     }
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Hl7Converter.kt:186]: Just a clean rename.

File: prime-router/src/main/kotlin/OrganizationService.kt [hunk 7]
```diff
@@ -1,16 +1,33 @@
...
+ * @param transport that the org wishes to recieve
  */
 data class OrganizationService(
     val name: String,
     val topic: String,
     val schema: String,
     val jurisdictionalFilter: Map<String, String> = emptyMap(),
     val transforms: Map<String, String> = emptyMap(),
+    val batch: Batch? = null,
     val address: String = "",
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/OrganizationService.kt:31]: Batch is the new set of properties that define the batching for the batch function.

File: prime-router/src/main/kotlin/azure/DatabaseAccess.kt [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+            config.jdbcUrl = databaseUrl
+            config.username = user
+            config.password = password
+            config.addDataSourceProperty("dataSourceClassName", "org.postgresql.ds.PGSimpleDataSource")
+            config.addDataSourceProperty("cachePrepStmts", "true")
+            config.addDataSourceProperty("prepStmtCacheSize", "250")
+            config.addDataSourceProperty("prepStmtCacheSqlLimit", "2048")
+            val dataSource = HikariDataSource(config)
+
+            val flyway = Flyway.configure().dataSource(dataSource).load()
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/azure/DatabaseAccess.kt:237]: Migration is quick when already done.

File: prime-router/src/main/kotlin/azure/WorkflowEngine.kt [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+
+/**
+ * Methods to add a new report to the workflow pipeline and to handle a step in the pipeline.
+ *
+ * @see gov.cdc.prime.router.Report
+ * @see QueueAccess
+ * @see DatabaseAccess.Header
+ */
+class WorkflowEngine(
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/azure/WorkflowEngine.kt:17]: This is the main coordinating engine. Replaces what was called ReportQueue.

File: prime-router/src/main/resources/log4j2.xml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
```
Comment (RickHawesUSDS) [inline prime-router/src/main/resources/log4j2.xml:2]: We didn't have this config file before, Log4j would complain.

File: prime-router/Dockerfile (hunk 1 above)
Comment (mry-usds) [reply prime-router/Dockerfile:12]: Cool - is much faster

File: prime-router/metadata/organizations.yml (hunk 3 above)
Comment (mry-usds) [reply prime-router/metadata/organizations.yml:27]: If possible - might want to switch to a cron like schema (e.g what if we want to send every 15 minutes or every hours between 0 and 8am?)

**Review by mry-usds — APPROVED** [review]
Fine with me - just minor question on the scheduling (can we use cron-like?)

**Comment by RickHawesUSDS** [comment]
@mry-usds I didn't think about cron expressions. Good addition.

**Pull request merged**